File: swara-core/src/main/java/com/swara/core/Composer.java
package com.swara.core;

import java.util.List;
import java.util.stream.Stream;

import com.swara.music.elements.Chord;
import com.swara.music.elements.Voice;

/**
 * A composer generates an original stream of {@link Chord}. Composers utilize a variety of machine
 * learning methods to create orginal, but representative music!
 */
@FunctionalInterface
public interface Composer {

    Stream<Chord> compose(List<Voice> examples);

}

File: swara-core/src/main/scala/com/swara/core/Composer.scala
package com.swara.core

import com.swara.music.elements.{Chord, Voice}

/**
 * A music composer. Composers generate original sequences of chords that are representative of but
 * distinct from a set of example voices. Composers may utilize a variety of machine learning and
 * statistical techniques to produce musically correct and interesting chord sequences.
 */
trait Composer {

  def compose(examples: Seq[Voice]): Iterator[Chord]

}
